pss/trojan/message.go

```diff
@@ -0,0 +1,215 @@
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The Swarm library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the Swarm library. If not, see <http://www.gnu.org/licenses/>.
+
+package trojan
+
+import (
+	"bytes"
+	"crypto/rand"
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"math/big"
+
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/ethersphere/swarm/chunk"
+	"github.com/ethersphere/swarm/storage"
+)
+
+// Topic is an alias for a 32 byte fixed-size array which contains an encoding of a message topic
+type Topic [32]byte
+
+// Message represents a trojan message, which is a message that will be hidden within a chunk payload as part of its data
+type Message struct {
+	length  [2]byte // big-endian encoding of Message payload length
+	topic   Topic
+	payload []byte
+	padding []byte
+}
+
+// MaxPayloadSize is the maximum allowed payload size for the Message type, in bytes
+const MaxPayloadSize = 4030
+
+var hashFunc = storage.MakeHashFunc(storage.BMTHash)
+
+var errPayloadTooBig = fmt.Errorf("message payload size cannot be greater than %d bytes", MaxPayloadSize)
+var errEmptyTargets = errors.New("target list cannot be empty")
+var errVarLenTargets = errors.New("target list cannot have targets of different length")
+
+// NewTopic creates a new Topic variable with the given input string
+// the input string is taken as a byte slice and hashed
+func NewTopic(topic string) Topic {
+	return Topic(crypto.Keccak256Hash([]byte(topic)))
+}
+
+// newMessage creates a new Message variable with the given topic and payload
+// it finds a length and nonce for the message according to the given input and maximum payload size
+func newMessage(topic Topic, payload []byte) (Message, error) {
+	if len(payload) > MaxPayloadSize {
+		return Message{}, errPayloadTooBig
+	}
+
+	// get length as array of 2 bytes
+	payloadSize := uint16(len(payload))
+	lengthBuf := make([]byte, 2)
+	binary.BigEndian.PutUint16(lengthBuf, payloadSize)
+
+	// set random bytes as padding
+	paddingLen := MaxPayloadSize - payloadSize
+	padding := make([]byte, paddingLen)
+	if _, err := rand.Read(padding); err != nil {
+		return Message{}, err
+	}
+
+	// create new Message var and set fields
+	m := new(Message)
+	copy(m.length[:], lengthBuf[:])
+	m.payload = payload
+	m.padding = padding
+
+	return *m, nil
+}
+
+// Wrap creates a new trojan chunk for the given targets and Message
+// a trojan chunk is a content-addressed chunk made up of span, a nonce, and a payload which contains the Message
+// the chunk address will have a matching prefix with one of the targets
+func Wrap(targets [][]byte, m Message) (chunk.Chunk, error) {
```

Comment:
also we may want a `context.Context` here passed on to `toChunk` re:// TODO: prevent infinite loop

Reply:
yeah, when you brought this up last time i wasn't sure whether you were asking to have a `Message` type as a receiver for this func, so i changed it back at that point (thought maybe you were just talking about the param type).

updated both `Wrap` and `toChunk` to use `Message` types as receivers.

since we are going to have new issues to work on polishing the mining, i will add the `context.Context` thing as a note in those issues. otherwise it's just going to be unused for now.